File: scripts/start-api.js
#!/usr/bin/env node
/**
 * YUVA API Server Starter (Node.js)
 * Cross-platform solution for starting the Python API server
 */

const { spawn } = require('child_process');
const path = require('path');
const fs = require('fs');

class APIServer {
  constructor() {
    this.process = null;
    this.isRunning = false;
  }

  start() {
    return new Promise((resolve, reject) => {
      // Check if Python is available
      const pythonCheck = spawn('python', ['--version'], { stdio: 'pipe' });
      
      pythonCheck.on('error', (error) => {
        reject(new Error(`Python not found: ${error.message}`));
      });

      pythonCheck.on('close', (code) => {
        if (code !== 0) {
          reject(new Error('Python not available'));
          return;
        }

        // Check if simple_api.py exists
        const apiPath = path.join(process.cwd(), 'simple_api.py');
        if (!fs.existsSync(apiPath)) {
          reject(new Error('simple_api.py not found'));
          return;
        }

        console.log('🚀 Starting YUVA API Server...');
        console.log('📍 API will be available at: http://localhost:5000');
        
        // Start the Python API server
        this.process = spawn('python', ['simple_api.py'], {
          stdio: 'inherit',
          cwd: process.cwd()
        });

        this.process.on('error', (error) => {
          reject(new Error(`Failed to start API server: ${error.message}`));
        });

        this.process.on('exit', (code) => {
          this.isRunning = false;
          if (code !== 0) {
            console.log(`❌ API server exited with code ${code}`);
          } else {
            console.log('✅ API server stopped gracefully');
          }
        });

        // Give the server a moment to start
        setTimeout(() => {
          this.isRunning = true;
          resolve();
        }, 2000);
      });
    });
  }

  stop() {
    if (this.process && this.isRunning) {
      console.log('🛑 Stopping API server...');
      this.process.kill('SIGTERM');
      this.isRunning = false;
    }
  }

  isServerRunning() {
    return this.isRunning;
  }
}

// If this script is run directly
if (require.main === module) {
  const apiServer = new APIServer();
  
  // Handle graceful shutdown
  process.on('SIGINT', () => {
    console.log('\n🛑 Shutting down...');
    apiServer.stop();
    process.exit(0);
  });

  process.on('SIGTERM', () => {
    apiServer.stop();
    process.exit(0);
  });

  // Start the server
  apiServer.start()
    .then(() => {
      console.log('✅ API server started successfully');
    })
    .catch((error) => {
      console.error('❌ Failed to start API server:', error.message);
      process.exit(1);
    });
}

module.exports = APIServer;

File: start_api.py
#!/usr/bin/env python3
"""
YUVA API Server Starter
A wrapper script that ensures the API server starts properly
"""

import subprocess
import sys
import time
import os
import signal
import threading
from pathlib import Path

def check_python():
    """Check if Python is available"""
    try:
        result = subprocess.run([sys.executable, '--version'], 
                              capture_output=True, text=True, timeout=5)
        if result.returncode == 0:
            print(f"✓ Python found: {result.stdout.strip()}")
            return True
    except Exception as e:
        print(f"❌ Python check failed: {e}")
    return False

def check_dependencies():
    """Check if required Python packages are available"""
    required_packages = ['urllib', 'json', 'http.server']
    missing_packages = []
    
    for package in required_packages:
        try:
            __import__(package)
        except ImportError:
            missing_packages.append(package)
    
    if missing_packages:
        print(f"❌ Missing required packages: {', '.join(missing_packages)}")
        return False
    
    print("✓ All required packages available")
    return True

def start_api_server():
    """Start the API server"""
    try:
        # Check if simple_api.py exists
        if not os.path.exists('simple_api.py'):
            print("❌ simple_api.py not found in current directory")
            return False
        
        print("🚀 Starting YUVA API Server...")
        print("📍 API will be available at: http://localhost:5000")
        print("🔄 Press Ctrl+C to stop the server")
        print("-" * 50)
        
        # Start the server
        process = subprocess.Popen([sys.executable, 'simple_api.py'], 
                                 stdout=subprocess.PIPE, 
                                 stderr=subprocess.STDOUT,
                                 universal_newlines=True,
                                 bufsize=1)
        
        # Stream output in real-time
        for line in iter(process.stdout.readline, ''):
            print(line.rstrip())
        
        process.wait()
        return process.returncode == 0
        
    except KeyboardInterrupt:
        print("\n🛑 Stopping API server...")
        if 'process' in locals():
            process.terminate()
        return True
    except Exception as e:
        print(f"❌ Failed to start API server: {e}")
        return False

def main():
    """Main function"""
    print("=" * 50)
    print("🎯 YUVA API Server Starter")
    print("=" * 50)
    
    # Check prerequisites
    if not check_python():
        print("❌ Python is not available or not working properly")
        sys.exit(1)
    
    if not check_dependencies():
        print("❌ Missing required dependencies")
        sys.exit(1)
    
    # Start the server
    success = start_api_server()
    
    if success:
        print("✅ API server stopped gracefully")
    else:
        print("❌ API server encountered an error")
        sys.exit(1)

if __name__ == "__main__":
    main()
